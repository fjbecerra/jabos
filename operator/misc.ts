import { ApiObjectProps, Chart } from 'cdk8s';
import { GitRepositoryProps } from './imports/jabos.io';
import { KubeJob, KubeJobProps, KubeRoleBinding } from './imports/k8s';
import { SyncRequest } from './metaControllerHooks';

export function k8sName(prefix: string, commit: string): string {
  return `${prefix.substring(0, 52)}-${commit}`.substring(0, 62);
}

export interface GitRepositoryPropsEx extends GitRepositoryProps {
  readonly status: {
    readonly latestCommit: string,
  }
}

export function getRepo(request: SyncRequest): GitRepositoryPropsEx {
  const namespace = request.object.metadata.namespace;
  const reposMap = (request.related || {})['GitRepository.jabos.io/v1'] /* as { [key: string]: GitRepositoryProps } */;
  var repo: GitRepositoryProps;

  // Have to filter, due to the following bug: https://github.com/metacontroller/metacontroller/issues/414
  const repos = Object.values(reposMap || {}).filter((r) => r.metadata?.namespace === namespace);
  if (repos.length > 0) {
    repo = repos[0] as unknown as GitRepositoryProps;
  }
  else {
    // Have to reject the sync to maintain idempotent response - due to: https://github.com/metacontroller/metacontroller/issues/414
    throw new Error(`No GitRepository from same namespace.`);
  }

  return {
    status: { latestCommit: '' },
    spec: { url: '', branch: '', ssh: null },
    ...repo,
  };
}

export function useExistingJob(chart: Chart, request: SyncRequest) {
  if (Object.values(request.attachments['Job.batch/v1']).length > 0) {
    var jobProps: KubeJobProps = Object.values(request.attachments['Job.batch/v1'])[0];
    return new KubeJob(chart, jobProps.metadata.name, jobProps);
  }
}

export function debugId(object: ApiObjectProps) {
  try {
    return `${object.metadata?.namespace}:${object.metadata?.name}`;
  }
  catch {
    return '(NO_OBJECT)';
  }
}

export function needNewBuild(request: SyncRequest): boolean {
  const object = request.object;
  const repo = getRepo(request);
  const existingJob = Object.values(request.attachments['Job.batch/v1']).length > 0;
  const builtCommit: string = (object.status || {}).builtCommit || '';
  const latestCommit: string = repo.status.latestCommit;
  return !existingJob && !!latestCommit && latestCommit !== builtCommit;
}
