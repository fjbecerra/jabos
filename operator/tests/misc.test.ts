import exp from 'constants';
import { debugId, getRepo } from '../misc';

test('misc getRepo no related', () => {
  expect(() => {
    getRepo({
      object: {
        metadata: {
          namespace: 'namespace_value'
        }
      },
      attachments: [],
      related: undefined,
      finalizing: false,
      controller: {}
    });
  }).toThrow('No GitRepository from same namespace.');
});

test('misc getRepo missing repo', () => {
  expect(() => {
    getRepo({
      object: {
        metadata: {
          namespace: 'namespace_value'
        }
      },
      attachments: [],
      related: {
        'GitRepository.jabos.io/v1': [
          undefined
        ]
      },
      finalizing: false,
      controller: {}
    });
  }).toThrow('No GitRepository from same namespace.');
});

test('misc getRepo missing metadata', () => {
  expect(() => {
    getRepo({
      object: {
        metadata: {
          namespace: 'namespace_value'
        }
      },
      attachments: [],
      related: {
        'GitRepository.jabos.io/v1': [
          {}
        ]
      },
      finalizing: false,
      controller: {}
    });
  }).toThrow('No GitRepository from same namespace.');
});

test('misc getRepo missing namespace', () => {
  expect(() => {
    getRepo({
      object: {
        metadata: {
          namespace: 'namespace_value'
        }
      },
      attachments: [],
      related: {
        'GitRepository.jabos.io/v1': [
          {
            metadata: {
            }
          }
        ]
      },
      finalizing: false,
      controller: {}
    });
  }).toThrow('No GitRepository from same namespace.');
});

test('misc getRepo bad namespace', () => {
  expect(() => {
    getRepo({
      object: {
        metadata: {
          namespace: 'namespace_value'
        }
      },
      attachments: [],
      related: {
        'GitRepository.jabos.io/v1': [
          {
            metadata: {
              namespace: 'bad_namespace_value'
            }
          }
        ]
      },
      finalizing: false,
      controller: {}
    });
  }).toThrow('No GitRepository from same namespace.');
});

test('misc getRepo no spec and status', () => {
  expect(getRepo({
    object: {
      metadata: {
        namespace: 'namespace_value'
      }
    },
    attachments: [],
    related: {
      'GitRepository.jabos.io/v1': [
        {
          metadata: {
            namespace: 'namespace_value'
          },
        }
      ]
    },
    finalizing: false,
    controller: {}
  })).toMatchSnapshot();
});

test('debugId no value', () => {
  expect(debugId({})).toMatchSnapshot();
});

test('debugId undefined', () => {
  expect(debugId(undefined)).toMatchSnapshot();
});
