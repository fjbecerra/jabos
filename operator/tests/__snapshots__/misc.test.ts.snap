// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`debugId no value 1`] = `"undefined:undefined"`;

exports[`debugId undefined 1`] = `"undefined:undefined"`;

exports[`misc getRepo no spec and status 1`] = `
Object {
  "apiVersion": "apiVersion_value",
  "kind": "kind_version",
  "metadata": Object {
    "namespace": "namespace_value",
  },
  "spec": Object {
    "branch": "",
    "ssh": null,
    "url": "",
  },
  "status": Object {
    "latestCommit": "",
  },
}
`;
